### PR TITLE
Encoding fixes, trying to add some high level optics.

### LINKED ISSUE #57: Uppercase HeXDigit's aren't interpreted correctly
```
> review _JCharEscaped (Hex (HexDigit4 HeXDigit0 HeXDigit0 HeXDigit0 HeXDigitA)) :: Char
'\NUL'
```

### PR DESCRIPTION
**Pull request by mankyKitty**
Tried to add some general `lens-aeson`-esque optics. Most of these cannot be made into law abiding optics due to the loss of information once the data structure is translated to a different Haskell type. In the process of adding tests to check that the `Prism'`s were lawful, this uncovered some bugs that this PR will contain the fixes for.

I also can't leave well enough alone, so... let's do this.

This PR should also fix #57 and #58 

This was partly due to a head->desk worthy implementation of a `JString -> Text` function that only went via an encoding that would strip anything but `ascii` or `latin1` then claim to be a UTF-8 format. My fault entirely and that should now be fixed, along with some extra tests to try to catch any regressions.

The other issue is that the ability to "BYO parsing package" is supported but is also the source of potential issues. As the original tests that were designed to catch this sort of failure would permit only half of a textual encoding/decoding process to be completed, leaving things in a broken state to be parsed "correctly" but not in with respect to the original inputs. The tests have been corrected but the flexibility will be improved to avoid this problem.

The story so far:

- Deleted some instances that would give rise to things that didn't obey the laws and introduced some bugs because of this.
- Deleted and renamed some functions.
- Fixed some documentation errors.
- Fixed a sneaky bug in the HeXDigit4 conversion function. There is more work to be done here as some values are not handled correctly.
- Added some prism law tests.
- Added some regression tests to confirm handling on some unicode values. Fixed some tests that should not have been passing.
- Added some files to use for testing to avoid possible issues from using literal `Text` or `ByteString` inputs.

### Fixes

- [x] Handling of `HeXDigit4` values is not correct. The bug is with the choice of optic, instead of producing a (type/failure) error when working with mixed-case hex values, it seems to be zero'ing them out.
- [x] Add regression tests

### Rework

This is partly to solve the confusion around how to even use the `ParseFn` and also to avoid future issues.
- [x] Redesign `ParseFn` to more capably handle:
    - [x] `Data.Text.Text`
    - [x] `Data.ByteString.ByteString`
- [x] Update documentation for `ParseFn` to match changes
- [x] Ruthlessly refactor or remove anything claiming to be an `Optic` if it does not satisfy the appropriate laws.
- [x] Update documentation for default parsing functions
- [x] Add default parsing functions using `attoparsec` as it's in dependency scope already and would give an easy ride for people that don't care to use a different package. Some cabal options might be required to allow people to exclude the package if so desired.

All the effort to convert from a `ByteString` builder to the `Text` builder may have been a red herring as the issue was simpler (I'm a numpty). It also should not be difficult to support both, so give users the option.
- [x] Generalise the `Builder` process to handle `Text` and `ByteString`
  - [x] Created a record type to hold the required functions for builders
  - [x] Create submodules to house the generalised builders (see `Waargonaut.Encode.Builder` and friends)
  - [x] Add test to ensure both builders produce identical output
- [x] Update documentation for `Encode` process to match changes

### Cleanup

- [x] Factor out components into more submodules. Some modules are overcrowded and the code is unpleasant. For starters (not exhaustive):
    - [x] `UnescapedJChar`
    - [x] `EscapedJChar`
    - [x] `HexDigit4`
    - [x] `Elem`
    - [x] `Elems`
    - [x] `JAssoc`
    - [x] `Decode.Runners`
- [x] Update documentation if required for module changes.
- [x] Delete commented out code
- [x] **_Stretch_**: Change all file textual encoding/decoding tests to `Test.Tasty.Golden` to avoid implementing my own slower, less informative, and buggy version.

### CODE REVIEW

File: src/Waargonaut/Types/JString.hs [hunk 1]
```diff
@@ -102,18 +100,16 @@ instance AsJString [JChar HeXDigit] where
...
 
 -- | Prism between a 'JString' and 'Text'.
 --
 -- JSON strings a wider range of encodings than 'Text' and to be consistent with
 -- the 'Text' type, these invalid types are replaced with a placeholder value.
 --
-_JStringText :: Prism' JString Text
-_JStringText = prism
-  (JString' . V.fromList . fmap utf8CharToJChar . Text.unpack)
-  (\x -> first (const x) . Text.decodeUtf8' . BS8.toStrict . BS8.pack . review _JString $ x)
```
Comment (mankyKitty) [inline src/Waargonaut/Types/JString.hs]: Upon reflection, this is the software development equivalent of: "I'll walk it through Mayan first"

File: src/Waargonaut.hs [hunk 2]
```diff
@@ -15,14 +21,126 @@ module Waargonaut
...
+{-# INLINE _Number #-}
+
+-- | Not a lawful 'Prism'' between some 'Json' and a 'Text' value
+_String :: Prism' Json Text
+_String = prism (E.asJson' E.text) (\j -> note j $ j ^? _JStr . _1 . _JStringText)
+{-# INLINE _String #-}
+
+-- | 'Prism'' between some 'Json' and a '()' value
+_Null :: Prism' Json ()
+_Null = prism (E.asJson' E.null) (\j -> note j $ () <$ j ^? _JNull)
```
Comment (tonymorris) [inline src/Waargonaut.hs]: `() <$`

can be

`void`

File: src/Waargonaut/Decode.hs [hunk 3]
```diff
@@ -582,7 +568,7 @@ withType
   -> JCurs
   -> DecodeResult f a
 withType t d c =
-  if maybe False (== t) $ jsonTypeAt (unJCurs c) then d c
+  if Just t == jsonTypeAt (unJCurs c) then d c
```
Comment (tonymorris) [inline src/Waargonaut/Decode.hs]: `elem`

Comment (mankyKitty) [reply src/Waargonaut/Decode.hs]: fixed

File: src/Waargonaut.hs (hunk 2 above)
Comment (mankyKitty) [reply src/Waargonaut.hs]: fixed

File: test/Types/Common.hs [hunk 4]
```diff
@@ -252,11 +232,30 @@ genWhitespace = Gen.element
...
+parseBS :: SD.Decoder Identity a -> ByteString -> Either (DecodeError, SD.CursorHistory) a
+parseBS d = SD.pureDecodeFromByteString AB.parseOnly d
+
+parseText :: SD.Decoder Identity a -> Text -> Either (DecodeError, SD.CursorHistory) a
+parseText d = SD.pureDecodeFromText AT.parseOnly d
+
+encodeJsonText :: Json -> Text
+encodeJsonText = TextL.toStrict . E.simplePureEncodeText E.json
+
+encodeText :: E.Encoder Identity a -> a -> TextL.Text
```
Comment (benkolera) [inline test/Types/Common.hs]: I still feel like encodeText/encodeBS and decodeText/decodeBS ought to be in Waargonaut. Make a opinionated choice about what the standard default should be but explain how they can peel a layer off and make a different choice if needed.  But then I guess that you have do do this for both the pure and non pure versions so it gets messy real quick. :(

File: test/json-data/goldens/backslash128.json.golden [hunk 5]
```diff
@@ -0,0 +1 @@
+{"a":"\\128"}
```
Comment (benkolera) [inline test/json-data/goldens/backslash128.json.golden]: Loooool. Our old friend. :)

File: test/Types/Common.hs (hunk 4 above)
Comment (mankyKitty) [reply test/Types/Common.hs]: I had scratched it off, but I think I will do this. It is starting to seem silly that you can't just get the package and start using it. `attoparsec` is in the gargantuan dependency tree anyway.

If I have these functions in a different module then it doesn't get too messy.

File: test/json-data/goldens/backslash128.json.golden (hunk 5 above)
Comment (mankyKitty) [reply test/json-data/goldens/backslash128.json.golden]: :rage:

File: src/Waargonaut.hs [hunk 6]
```diff
@@ -141,44 +259,59 @@ import           Waargonaut.Types.Json (JType (..), Json (..), parseWaargonaut,
...
+-- functions provided for some of the common input text-like types.:
 --
 -- @
--- runDecode
---   :: Monad f
---   => Decoder f a
---   -> ParseFn
---   -> JCurs
---   -> f (Either (DecodeError, CursorHistory) a)
+-- decodeFromByteString
```
[the remark this answers is not in the document]
Comment (mankyKitty) [reply src/Waargonaut.hs]: :tada:

**Pull request merged**